src/server/pkg/sync/sync.go

```diff
@@ -38,6 +39,75 @@ func NewPuller() *Puller {
 	}
 }
 
```

Comment:
Ahh, I see. What changed is that I added a `PullDiff` method which pulls a different view of the filesystem based on the diff rather than the full content of the directory. Also as part of this I refactored some shared logic between `Pull` and `PullDiff` into helper functions.